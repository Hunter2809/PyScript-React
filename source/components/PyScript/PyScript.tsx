import {
  createElement,
  FC,
  PropsWithChildren,
  ReactHTML,
  useMemo,
} from "react";

export type PyScriptProperties = {
  children: string;
  output?: string;
  generateOutputTag?: boolean | keyof ReactHTML;
  pyEnvContent?: string | string[];
};

const PyScript: FC<PropsWithChildren<PyScriptProperties>> = ({
  children,
  output,
  generateOutputTag,
  pyEnvContent,
}: PyScriptProperties): JSX.Element => {
  const pyEnvFixedContent = useMemo(() => {
    return Array.isArray(pyEnvContent)
      ? pyEnvContent
          .map((element) => {
            return `- ${element};`;
          })
          .join(" ")
      : pyEnvContent;
  }, [pyEnvContent]);
  return (
    <>
      {pyEnvFixedContent && <py-env>{pyEnvFixedContent}</py-env>}
      {generateOutputTag &&
        createElement(
          typeof generateOutputTag === "string" ? generateOutputTag : "div",
        )}
      <py-script output={output}>{children}</py-script>
    </>
  );
};

export default PyScript;
