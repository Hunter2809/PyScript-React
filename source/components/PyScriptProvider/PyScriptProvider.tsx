import { FC, memo, PropsWithChildren } from "react";
import { Helmet, HelmetProvider } from "react-helmet-async";

export type PyScriptProviderProperties = {
  cssSource?: string;
  jsSource?: string;
};

const PyScriptProvider: FC<PropsWithChildren<PyScriptProviderProperties>> = ({
  cssSource = "https://pyscript.net/alpha/pyscript.css",
  jsSource = "https://pyscript.net/alpha/pyscript.js",
  children,
}) => {
  return (
    <HelmetProvider>
      <Helmet>
        <link rel="stylesheet" href={cssSource} />
        <script defer src={jsSource}></script>
      </Helmet>
      <div></div>
      {children}
    </HelmetProvider>
  );
};

export default memo(PyScriptProvider);
