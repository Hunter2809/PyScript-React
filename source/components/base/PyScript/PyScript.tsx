import propTypes from "prop-types";
import {
  createElement,
  DetailedHTMLProps,
  FC,
  HTMLAttributes,
  ReactHTML,
  WeakValidationMap,
} from "react";
import PyEnv, { PyEnvProperties } from "~components/base/PyEnv/PyEnv";
import ReactElementProps from "~types/ReactElementProps/ReactElementProps";

export type PyScriptPropertiesBase = Omit<
  ReactElementProps<
    DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>
  >,
  "children"
> & {
  children: string;
};

export type PyScriptPropertiesWitOutputPart = {
  output: string;
  generateOutputTag?: boolean | keyof ReactHTML;
};

export type PyScriptPropertiesWitoutOutputPart = {
  output?: never;
  generateOutputTag?: never;
};

export type PyScriptPropertiesWithPyEnvPart = {
  pyEnvContent: PyEnvProperties["children"];
  pyEnvProps?: Omit<PyEnvProperties, "children">;
};

export type PyScriptPropertiesWithWithoutPyEnvPart = {
  pyEnvContent?: never;
  pyEnvProps?: never;
};

export type PyScriptProperties =
  | PyScriptPropertiesBase &
      (
        | PyScriptPropertiesWithPyEnvPart
        | PyScriptPropertiesWithWithoutPyEnvPart
      ) &
      (PyScriptPropertiesWitOutputPart | PyScriptPropertiesWitoutOutputPart);

const PyScript: FC<PyScriptProperties> = ({
  children,
  output,
  generateOutputTag,
  pyEnvContent,
  pyEnvProps,
  ...rest
}: PyScriptProperties): JSX.Element => {
  return (
    <>
      {pyEnvContent && <PyEnv {...pyEnvProps}>{pyEnvContent}</PyEnv>}
      {output &&
        generateOutputTag &&
        createElement(
          typeof generateOutputTag === "string" ? generateOutputTag : "div",
          {
            id: output,
          },
        )}
      <py-script {...rest} output={output}>
        {children}
      </py-script>
    </>
  );
};

PyScript.propTypes = {
  children: propTypes.string.isRequired,
  output: propTypes.string,
  generateOutputTag: propTypes.oneOfType([propTypes.string, propTypes.bool]),
  pyEnvContent: propTypes.oneOfType([propTypes.string, propTypes.array]),
  pyEnvProps: propTypes.object,
} as WeakValidationMap<PyScriptProperties>;

export default PyScript;
