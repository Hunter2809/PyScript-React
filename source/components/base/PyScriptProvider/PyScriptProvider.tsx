import propTypes from "prop-types";
import { FC, memo, PropsWithChildren } from "react";
import { Helmet, HelmetProvider } from "react-helmet-async";

export type PyScriptProviderProperties = PropsWithChildren<{
  cssSource?: string;
  jsSource?: string;
}>;

const PyScriptProvider: FC<PyScriptProviderProperties> = ({
  cssSource = "https://pyscript.net/alpha/pyscript.css",
  jsSource = "https://pyscript.net/alpha/pyscript.js",
  children,
}: PyScriptProviderProperties): JSX.Element => {
  return (
    <HelmetProvider>
      <Helmet>
        <link rel="stylesheet" href={cssSource} />
        <script defer src={jsSource}></script>
      </Helmet>
      {children}
    </HelmetProvider>
  );
};

PyScriptProvider.propTypes = {
  cssSource: propTypes.string,
  jsSource: propTypes.string,
  children: propTypes.oneOfType([
    propTypes.arrayOf(propTypes.node),
    propTypes.node,
  ]).isRequired,
};

export default memo(PyScriptProvider);
