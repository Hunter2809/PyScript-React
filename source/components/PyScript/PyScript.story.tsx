import { ComponentMeta, ComponentStory } from "@storybook/react";
import PyScript, { PyScriptProperties } from "./PyScript";
import { FunctionComponent } from "react";
import PyScriptProvider from "../PyScriptProvider/PyScriptProvider";

export default {
  title: "PyScript",
  component: PyScript,
} as ComponentMeta<typeof PyScript>;

const Template: ComponentStory<FunctionComponent<PyScriptProperties>> = ({
  ...rest
}: PyScriptProperties): JSX.Element => {
  return (
    <PyScriptProvider>
      <PyScript {...rest} />
    </PyScriptProvider>
  );
};

export const PyScriptExample: ComponentStory<typeof PyScript> = Template.bind(
  {},
);

PyScriptExample.args = {
  children: 'print("hello")',
  output: "folium",
  generateOutputTag: true,
  pyEnvContent: ["folium"],
};
