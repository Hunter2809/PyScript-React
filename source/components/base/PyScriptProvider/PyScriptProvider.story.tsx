import { ComponentMeta, ComponentStory } from "@storybook/react";
import { FunctionComponent } from "react";
import PyScriptProvider, {
  PyScriptProviderProperties,
} from "./PyScriptProvider";

export default {
  title: "PyScriptProvider",
  component: PyScriptProvider,
} as ComponentMeta<typeof PyScriptProvider>;

const Template: ComponentStory<
  FunctionComponent<PyScriptProviderProperties>
> = ({ ...rest }: PyScriptProviderProperties): JSX.Element => {
  return (
    <PyScriptProvider {...rest}>
      <py-script>print("Hello World!")</py-script>
    </PyScriptProvider>
  );
};

export const PyScriptProviderExample: ComponentStory<typeof PyScriptProvider> =
  Template.bind({});

PyScriptProviderExample.args = {
  cssSrc: "https://pyscript.net/alpha/pyscript.css",
  jsSrc: "https://pyscript.net/alpha/pyscript.js",
};
