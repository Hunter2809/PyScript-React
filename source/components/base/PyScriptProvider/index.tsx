import PyScriptProvider from "~components/base/PyScriptProvider/PyScriptProvider";

export default PyScriptProvider;
