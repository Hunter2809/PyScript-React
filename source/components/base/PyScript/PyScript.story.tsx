import { ComponentMeta, ComponentStory } from "@storybook/react";
import PyScript, { PyScriptProperties } from "./PyScript";
import { FunctionComponent } from "react";
import PyScriptProvider from "../PyScriptProvider/PyScriptProvider";

export default {
  title: "PyScript",
  component: PyScript,
} as ComponentMeta<typeof PyScript>;

const Template: ComponentStory<FunctionComponent<PyScriptProperties>> = ({
  ...rest
}: PyScriptProperties): JSX.Element => {
  return (
    <PyScriptProvider>
      <PyScript {...rest} />
    </PyScriptProvider>
  );
};

export const PyScriptExample: ComponentStory<typeof PyScript> = Template.bind(
  {},
);

PyScriptExample.args = {
  children: 'print("hello")',
  output: "folium",
  generateOutputTag: true,
  pyEnvContent: ["folium"],
};

export const PyScriptExampleFibonacci: ComponentStory<typeof PyScript> =
  Template.bind({});

PyScriptExampleFibonacci.args = {
  children: `
  class Fibonacci:
    def __init__(self):
        self.fibonacci_cache = {}
    def sequence(self, n) -> int:
        if n in self.fibonacci_cache:
            return self.fibonacci_cache[n]
        if n == 1:
            return 1
        elif n == 2:
            return 1
        elif n > 2:
            self.fibonacci_cache[n] = self.sequence(n-1) + self.sequence(n-2)
            return self.fibonacci_cache[n]


  fibonacci = Fibonacci()
  for x in range(100):
    print(fibonacci.sequence(x))

  
  `,
};
