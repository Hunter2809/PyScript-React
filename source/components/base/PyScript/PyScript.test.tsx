import { cleanup, render, RenderResult } from "@testing-library/react";
import { FC } from "react";
import { expectType } from "tsd";
import PyScript, {
  PyScriptProperties,
} from "~components/base/PyScript/PyScript";
import PyScriptProvider from "~components/base/PyScriptProvider/PyScriptProvider";

afterEach(cleanup);

const PyScriptHelloWorldComponent: FC = (): JSX.Element => {
  return (
    <PyScriptProvider>
      <PyScript data-testid="py-script">print("Hello world!")</PyScript>
    </PyScriptProvider>
  );
};
const PyScriptFoliumMapComponent: FC = (): JSX.Element => {
  return (
    <PyScriptProvider>
      <PyScript output="folium" generateOutputTag pyEnvContent={["folium"]}>
        {`
from folium import Map
variable = Map(location=[45.5236, -122.6750])
variable
`}
      </PyScript>
    </PyScriptProvider>
  );
};

describe("PyScript", (): void => {
  it("Renders component properly", async (): Promise<void> => {
    const { container }: RenderResult = render(<PyScriptHelloWorldComponent />);
    expect(container).toMatchSnapshot();
  });
  it("Compile component with Python properly", (): void => {
    const { getByTestId, debug }: RenderResult = render(
      <PyScriptHelloWorldComponent />,
    );
    const pyScript: HTMLElement = getByTestId("py-script");
    setTimeout((): void => {
      debug();
      expect(pyScript).toContainEqual("Hello world!");
    }, 5_000);
  });
  it("Renders component properly with all props", async (): Promise<void> => {
    const { container }: RenderResult = render(<PyScriptFoliumMapComponent />);
    expect(container).toMatchSnapshot();
  });
  it("Has correct type", (): void => {
    expectType<FC<PyScriptProperties>>(PyScript);
  });
});
