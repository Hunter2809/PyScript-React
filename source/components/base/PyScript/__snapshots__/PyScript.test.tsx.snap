// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PyScript Renders component properly 1`] = `
<div>
  <py-script
    data-testid="py-script"
  >
    print("Hello world!")
  </py-script>
</div>
`;

exports[`PyScript Renders component properly with all props 1`] = `
<div>
  <py-env>
    - folium
  </py-env>
  <div
    id="folium"
  />
  <py-script
    output="folium"
  >
    
from folium import Map
variable = Map(location=[45.5236, -122.6750])
variable

  </py-script>
</div>
`;
